Support for using buildpackages in a CNB Registry

jkutner opened the pull request:
This PR adds support for using buildpackages from a [CNB Registry](https://github.com/buildpacks/rfcs/blob/master/text/0022-client-side-buildpack-registry.md).

With `build` this looks like:

```sh-session
$ pack build --buildpack urn:cnb:registry:example/foo myimage
```

With `create-builder`, the `builder.toml` might include:

```toml
[[buildpacks]]
  image = "urn:cnb:registry:example/foo"
```

In both cases, the registry location can be specified using the `-r` option. For example:

```sh-session
$ pack create-builder mybuilder -r https://github.com/buildpacks/registry -b builder.toml
```

jkutner commented on internal/commands/create_builder.go:
```diff
@@ -49,6 +53,7 @@ func CreateBuilder(logger logging.Logger, client PackClient) *cobra.Command {
 	}
 	cmd.Flags().BoolVar(&flags.NoPull, "no-pull", false, "Skip pulling build image before use")
 	cmd.Flags().StringVarP(&flags.BuilderTomlPath, "builder-config", "b", "", "Path to builder TOML file (required)")
+	cmd.Flags().StringVarP(&flags.Registry, "buildpack-registry", "r", cfg.DefaultRegistry, "Buildpack Registry URL")
```
we don't yet. that will come later

jkutner commented on testhelpers/testhelpers.go:
```diff
@@ -580,3 +580,37 @@ func writeTAR(t *testing.T, srcDir, tarDir string, mode int64, w io.Writer) {
 	AssertNil(t, err)
 }
+
+func RecursiveCopyNow(t *testing.T, src, dst string) {
```
yes, otherwise Git sees it as a diff

jkutner commented on internal/buildpack/locatortype.go:
```diff
@@ -49,6 +51,13 @@ func GetLocatorType(locator string, buildpacksFromBuilder []dist.BuildpackInfo)
+	if strings.HasPrefix(locator, fromRegistryPrefix+":") {
+		// if !registryMatchFound(locator, buildpacksFromBuilder) {
+		// 	return InvalidLocator, fmt.Errorf("%s is not a valid identifier", style.Symbol(locator))
+		// }
```
i think this will all get refactored when we implement the Buildpack URIs RFC

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+func (r *Cache) createCache() error {
+	root, err := ioutil.TempDir("", "registry")
```
so that we can rename it. The go-git api doesn't let us clone directly to an arbitrary dir name (at least a way that I could figure out).

jkutner replied:
and I didn't want to in the PACK_HOME and have it collide with something

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+
+type Cache struct {
+	URL  string
+	Root string
```
fixed

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+	if _, err := ggcrname.ParseReference(b.Address, ggcrname.WeakValidation); err != nil {
+		return err
+	}
+	_, err := ggcrname.NewDigest(b.Address)
```
I don't follow. What are you saying?

jromero replied:
Sorry it was a note that I intended to expand on during the review process. I got it's purpose after looking at the implementation of the method. 👍

jkutner commented on internal/registry/registry_test.go:
```diff
@@ -0,0 +1,175 @@
+					Author: &object.Signature{
+						Name:  "John Doe",
+						Email: "john@doe.org",
+						When:  time.Now(),
```
I can't remember the exact result, but the zeroed out timestamps broke Git

jromero replied:
Ok, it does look like it's limited to the tests only and nothing we should be concerned with in the standard user flow correct?

jkutner replied:
yea, correct. This never happens in the containers, so the git repo never have it's timestamps zeroed out.

jkutner commented on internal/registry/registry_test.go:
```diff
@@ -0,0 +1,175 @@
+	// copy fixture to temp dir
+	registryFixtureCopy := filepath.Join(tmpDir, "registryCopy")
+
+	h.RecursiveCopyNow(t, filepath.Join("..", "..", "testdata", "registry"), registryFixtureCopy)
```
I can't remember the exact result, but the zeroed out timestamps broke Git

jkutner commented on internal/registry/registry_test.go:
```diff
@@ -0,0 +1,175 @@
+	spec.Run(t, "Buildpack", func(t *testing.T, when spec.G, it spec.S) {
+		when("#Validate", func() {
+			it("errors when address is missing", func() {
+				b := registry.Buildpack{
```
I don't have a constructor because this comes from de-serialized JSON. But i have moved the `Validate` call into `LocateBuildpack` so that it never needs to be called elsewhere.

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+	Root string
+}
+
+func NewRegistryCache(home, registryURL string) (Cache, error) {
```
fixed

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+	return err
+}
+
+func (r *Cache) readEntry(ns, name, version string) (Entry, error) {
```
fixed

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+		return errors.Wrap(err, "could not access registry cache")
+	}
+
+	if len(remotes) != 1 || len(remotes[0].Config().URLs) != 1 {
```
yea, I was thinking this was one of many ways we could roughly check to make sure the repo has been dorked with, or corrupted. idk, I can drop it

jromero replied:
I think we should remove it, just incase. There seems to be no value in adding that restriction. 

We should replace it with an explicit check to ensure `origin` remote exists though.

jkutner replied:
fixed by 2e4256d

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+
+func NewRegistryCache(home, registryURL string) (Cache, error) {
+	if _, err := os.Stat(home); err != nil {
+		return Cache{}, err
```
Yea, looks like it only gets created when we `Write` the config file in PACK_HOME. We should definitely create PACK_HOME if it doesn't exist, but I'm not sure _where_ that should happen.

@jromero I don't have a strong preference. Want to tell me where to put the `MkdirAll` call?

jromero replied:
Based on the pattern in this repo it should be in `common.go` -> `getRegistry` after `config.PackHome()` is called. Make sure to call `config.MkdirAll`. Not particularly fond of this pattern but we can improve it later.

jkutner replied:
Fixed in 31864bc

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+	if b.Address == "" {
+		return errors.New("address is a required field")
+	}
+	if _, err := ggcrname.ParseReference(b.Address, ggcrname.WeakValidation); err != nil {
```
1. yea, but I'm not really sure how to do that in Golang. Like this will be an instance of a `Digest` in ggcr, but we only see it as a `Reference`.
2. tag and digest are mutually exclusive, so the call to `NewDigest` will error if there's a tag IIRC

natalieparellano replied:
Is https://github.com/google/go-containerregistry/pull/703/files#diff-384cbc32f17b860e5335ef202256ac9eR69 relevant to this discussion?

jkutner replied:
actually, all `ParseReference` does is call `NewDigest`, so I think i can just omit that call. https://github.com/google/go-containerregistry/blob/master/pkg/name/ref.go#L44

jkutner commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,243 @@
+	case len(name) < 3:
+		indexDir = name
+	case len(name) < 4:
+		indexDir = filepath.Join(name[:2], name[2:])
```
fixed

elbandito commented on internal/registry/registry.go:
```diff
@@ -0,0 +1,250 @@
+
+	key := sha256.New()
+	key.Write([]byte(normalizedURL.String()))
+	cacheDir := fmt.Sprintf("%s-%s", defaultRegistryDir, hex.EncodeToString(key.Sum(nil)))
```
@jkutner I might be misunderstanding how the registry cache works, but it appears every time we enter this logic flow we're creating a new unique cache.  Is this correct?  How will it find a previous registry cache with `hex.EncodeToString(key.Sum(nil))` resulting in a unique string?

jkutner replied:
As long as the URL for the registry is the same, the registry directory will be the same. In this way, you can have multiple registry caches in your PACK_HOME at once (maybe a cache for heroku-registry, and a cache for buildpacks-io-registry or whatever)

jkutner commented:
@jromero i'm still trying to get windows to pass on CI, but otherwise this is ready for another look. I also have some questions on a few of your comments

jkutner commented:
@jromero thanks!

> The --buildpack-registry flag doesn't tell me what the default is.

I don't want to advertise the default yet since this is still like an early beta behind a flag (and the URL might change). Can we hold off on broadcasting this until the registry is established?

> I would recommend that we use a capital short flag -R.

Fixed in 579ba2e 

> Could we make the error message clearer? What authentication is it referring to?

Fixed in fb930ea 

> The log, in verbose, should print what registry it's pulling a given buildpack from.

I added this in 588c0ee8. We could add a lot more logging, so let me know if there's anything specific you think needs to be in there. I also logged just the host/path to avoid logging secrets if they're embedded in the URL.

> A URN is a URI, I would expect URNs to be recognized in the uri key.

Fixed in 0fa5b88
